src/connectors/impls/psql_repl/postgres_replication/serializer.rs

```diff
@@ -0,0 +1,347 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use postgres_protocol::message::backend::{
+    Column, LogicalReplicationMessage, ReplicaIdentity, Tuple, TupleData, XLogDataBody,
+};
+use serde::ser::{Error, SerializeStruct};
+use serde::{Serialize, Serializer};
+use std::fmt;
+
+/// `SerializedTuple`: A wrapper struct around a Tuple object that provides a Serialize implementation
+/// for it. This struct is used to serialize the tuple data of a `PostgreSQL` message into JSON format.
+
+pub(crate) struct SerializedTuple<'a>(pub &'a Tuple);
+
+impl<'a> Serialize for SerializedTuple<'a> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let data = self
+            .0
+            .tuple_data()
+            .iter()
+            .map(SerializedTupleData::from_tuple_data)
+            .collect::<Vec<_>>();
+        let mut state = serializer.serialize_struct("Tuple", 1)?;
+        state.serialize_field("data", &data)?;
+        state.end()
+    }
+}
+
+/// `SerializedOptionTuple`: Similar to `SerializedTuple`, but for an optional Tuple object.
+/// If the inner Option is Some, this struct serializes the tuple data. If the inner Option is None,
+/// this struct serializes None.
+pub(crate) struct SerializedOptionTuple<'a>(pub Option<&'a Tuple>);
+
+impl<'a> Serialize for SerializedOptionTuple<'a> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match self.0 {
+            Some(tuple) => {
+                let data = tuple
+                    .tuple_data()
+                    .iter()
+                    .map(SerializedTupleData::from_tuple_data)
+                    .collect::<Vec<_>>();
+                let mut state = serializer.serialize_struct("Tuple", 1)?;
+                state.serialize_field("data", &data)?;
+                state.end()
+            }
+            None => serializer.serialize_none(),
+        }
+    }
+}
+
+impl<'a, 'b> From<&'b Tuple> for SerializedTuple<'a>
+where
+    'b: 'a,
+{
+    fn from(tuple: &'b Tuple) -> Self {
+        SerializedTuple(tuple)
+    }
+}
+
+/// `SerializedTupleData` : An enum that represents the different types of data that can be present in
+/// a tuple. This struct is used by `SerializedTuple` and `SerializedOptionTuple`
+/// to serialize tuple data into JSON format.
+#[derive(Serialize)]
+#[serde(untagged)]
+enum SerializedTupleData {
+    Null,
+    UnchangedToast,
+    Text(String),
```

Comment:
I think to some degree we get type information from PG, this is not a concern in this second but we might want to later decode the type into a more correct form of internal representation. the Clickhouse connector does some of that :)